rasa/utils/tensorflow/layers.py

```diff
@@ -685,24 +697,49 @@ def _sample_negatives(
             labels_bad_negs,
         )
 
-    @staticmethod
-    def sim(a: tf.Tensor, b: tf.Tensor, mask: Optional[tf.Tensor] = None) -> tf.Tensor:
+    def sim(
+        self, a: tf.Tensor, b: tf.Tensor, mask: Optional[tf.Tensor] = None
+    ) -> tf.Tensor:
         """Calculate similarity between given tensors."""
-
+        if self.similarity_type == COSINE:
```

Comment:
for `margin` loss, it seems to me you'd apply double normalization during prediction, I think it'll be clearer if you make `similarity_type` an input argument to this method

Reply:
Adding as an input to the function would mean we pass the value of the class's own instance variable `similarity_type` to its own function. It seemed very weird to do that and moreover, we will have to make sure that we pass the correct value in all parts where `.sim` is called.
I think it's fine that we double normalize in this case because the output would still be correct and refactoring it for `margin` loss seems an overkill as we will get rid of it eventually.
Once we get rid of margin loss we should do a clean up since there will only be one similarity type `inner` during training and different types of similarity during inference which  should be cleaner.
Wdyt?

Reply:
maybe then moving normalization to `margin` loss?

Reply:
In that case we will have to restrict `cosine` similarity only to `margin` loss. Currently we just raise a warning when `similarity_type=cosine` and `loss_type=cross_entropy/softmax` is used.
I guess we can't hard deprecate the combination in a minor?

Reply:
aaah let's leave it then. Could you create an issue about deprecating `margin` loss? Maybe some of the users will chime in

Reply:
Done - https://github.com/RasaHQ/rasa/issues/7891